### PR TITLE
feat(tcp tls): Added sni property to conn object.

### PR DESCRIPTION
**Pull request by ihack2712**
Added a `sni` property to the `Conn` object. The `sni` property will be `undefined` if the SNI wasn't included in the TLS handshake, or if the TLS protocol wasn't used at all.

The `sni` property will be passed by `op_accept_tls`.

### CODE REVIEW

**Comment by ihack2712** [comment]
This resolves #9109 btw.

**Review by ry — COMMENTED** [review]
Thanks - nice to have this feature but it needs a test.

**Comment by ihack2712** [comment]
Added a test, was working locally. Excited to see if it passes the tests here :3

Anyhow, I'm in a little need of assistance regarding adding the feature as unstable.

File: cli/tests/unit/tls_test.ts [hunk 1]
```diff
@@ -232,3 +233,41 @@ unitTest(
...
+    async function connect() {
+      const conn = await Deno.connectTls({
+        hostname,
+        port,
+        certFile: rootCertFile,
+      });
+      conn.close();
+    }
+
+    await resolvable;
```
Comment (lucacasonato) [inline cli/tests/unit/tls_test.ts:271]: I think this test can be written without a `deferred`.

Also it looks like the `connect` function is never called right? It looks to me like this test should just be hanging forever.

```suggestion
  async function getAndConnectWithSNI(): Promise<void> {
    const hostname = "localhost";
    const port = 3600;
    const certFile = "cli/tests/tls/localhost.crt";
    const keyFile = "cli/tests/tls/localhost.key";
    const rootCertFile = "cli/tests/tls/RootCa.pem";

    const listener = Deno.listenTls({
      hostname,
      port,
      certFile,
      keyFile,
    });

    const incomingPromise = listener.accept();
    
    const outgoing = await Deno.connectTls({
      hostname,
      port,
      certFile: rootCertFile,
    });
    outgoing.close();
    
    const incoming = await incomingPromise;
    assertEquals((incoming as Deno.Conn & { sni?: string }).sni, hostname);
    incoming.close();
    listener.close();
```

File: cli/tests/unit/tls_test.ts [hunk 2]
```diff
@@ -232,3 +233,41 @@ unitTest(
...
+
+    const listener = Deno.listenTls({
+      hostname,
+      port,
+      certFile,
+      keyFile,
+    });
+
+    listener.accept().then((conn) => {
+      assertEquals((conn as Deno.Conn & { sni?: string }).sni, hostname);
```
Comment (lucacasonato) [inline cli/tests/unit/tls_test.ts:254]: You need to fix these typings. Please add the correct typings to the `cli/dts/lib.deno.unstable.d.ts` file.

File: runtime/ops/tls.rs [hunk 3]
```diff
@@ -398,6 +398,13 @@ async fn op_accept_tls(
     .try_or_cancel(cancel)
     .await?;
 
+  let sni = tls_stream
+    .get_ref()
+    .1
+    .get_sni_hostname()
+    .unwrap()
```
Comment (lucacasonato) [inline runtime/ops/tls.rs:405]: You can't unwrap here. TLS connections do not always have the SNI set. SNI must be an `Option`.

File: cli/tests/unit/tls_test.ts (hunk 2 above)
Comment (ihack2712) [reply cli/tests/unit/tls_test.ts:254]: I tried adding the typings to `lib.deno.unstable.d.ts`. However, during the tests they were not applied.

File: cli/tests/unit/tls_test.ts (hunk 1 above)
Comment (ihack2712) [reply cli/tests/unit/tls_test.ts:271]: Yes, you're right! Don't know why my changes weren't pushed, the tests worked locally before pushing. I'm digging into it.

File: runtime/ops/tls.rs (hunk 3 above)
Comment (ihack2712) [reply runtime/ops/tls.rs:405]: Fixed in next push thanks to @SyrupThinker!

File: cli/dts/lib.deno.unstable.d.ts [hunk 4]
```diff
@@ -1255,6 +1255,24 @@ declare namespace Deno {
...
+    readonly remoteAddr: Addr;
+    /** The resource ID of the connection. */
+    readonly rid: number;
+    /** **UNSTABLE**: yet to be vetted.
+     * The SNI the client sent to the server during the TLS handshake. If the
+     * connection was not made over TLS, or the client did not send the SNI
+     * during the [client hello](https://tools.ietf.org/html/rfc5246#section-7.4.1.2)
+     * message this property will be undefined.
+     */
+    readonly sni?: string;
```
Comment (lucacasonato) [inline cli/dts/lib.deno.unstable.d.ts:1271]: Typing is not correct. It should be `string | null` not `string | undefined`.

File: runtime/ops/tls.rs [hunk 5]
```diff
@@ -398,6 +398,12 @@ async fn op_accept_tls(
     .try_or_cancel(cancel)
     .await?;
 
+  let sni = tls_stream
```
Comment (lucacasonato) [inline runtime/ops/tls.rs:401]: Wrap this in an if statement to check if unstable is enabled. If it is not, sni should be None.

**Comment by piscisaureus** [comment]
@piscisaureus

**Comment by lucacasonato** [comment]
@ihack2712 Thanks for the PR, but we are not going to land the PR as the proposed API has no path to stabilisation, as exposing the SNI (and ALPN and friends) should be done as a more general redesign of the TLS API to support dynamic certificates.

**Comment by ihack2712** [comment]
Okay! You mention dynamic certificates, is that on the roadmap?

**Pull request closed without merging**